### ISSUE TITLE
Adjust line inset value for lines with `line-gap-width` in shader?

### ISSUE THREAD

**Issue by nickidlugash** [mapbox/mapbox-gl-js#4708]
### Steps to Trigger Behavior

 1. Create a line style layer
 2. Create another line style layer with a `line-gap-width` value that equals the `line-width` value of the first line style layer

### Expected Behavior

There is no visible gap between the two lines

### Actual Behavior

There is a small visible gap between the two lines:

<img width="610" alt="screen shot 2017-05-11 at 10 12 39 pm" src="https://cloud.githubusercontent.com/assets/5607844/25979515/fb16a8b2-3696-11e7-9d89-a8005aeb8557.png">

The current shader behavior is to adjust the inset value of lines with a gapwidth to account for the antialiasing, but I think that this is causing the visible gap.

If you replace [this line](https://github.com/mapbox/mapbox-gl-js/blob/master/src/shaders/line.vertex.glsl#L58) in the line vertex shader:
`float inset = gapwidth + (gapwidth > 0.0 ? ANTIALIASING : 0.0);`

with

`float inset = gapwidth;`

There is no longer a visible gap. This does make lines a bit visibly thicker when there isn't a corresponding line whose `line-width` equals its `line-gap-width` drawn on top of it. However, I think the main use case of the `line-gap-width` property is to use it with another line drawn on top of it, so I think it may be better to optimize for that use case. If there's a solution that would be optimized for both, that would be even better.

Here's what the current behavior looks like:

<img width="870" alt="line-gap-width-current" src="https://cloud.githubusercontent.com/assets/5607844/25979459/92558988-3696-11e7-8baa-7999245d4173.png">

- Green lines have `line-width: A`
- Double maroon lines drawn under green lines have `line-gap-width: A` and `line-width: B`
- Double maroon lines not under green lines also have `line-gap-width: A` and `line-width: B`, for comparison
- Single maroon lines have `line-width: B`, for comparison

Here's what the altered behavior looks like:

<img width="880" alt="line-gap-width-no-visible-gap" src="https://cloud.githubusercontent.com/assets/5607844/25979820/ab2f433e-3698-11e7-81e3-85c97804f827.png">

We've used `line-gap-width` extensively in our core styles for a while, for roads, but I think the color contrast between the casing and the background color has always been so low that this issue wasn't really visible. We've recently been testing more high-contrast styling, and we are now finding that this is an issue. (I tested older version of mapbox-gl-js, I think this has probably always been consistent behavior).

Does this seem like something we can adjust?

/cc @aparlato

**Comment by nickidlugash** [comment]
Actually, also adjusting [this line](https://github.com/mapbox/mapbox-gl-js/blob/master/src/shaders/line.vertex.glsl#L59) from 

`float outset = gapwidth + width * (gapwidth > 0.0 ? 2.0 : 1.0) + ANTIALIASING;`

to

`float outset = gapwidth + width * (gapwidth > 0.0 ? 2.0 : 1.0) + (gapwidth > 0.0 ? 0.0 : ANTIALIASING);`

I think fixes the width discrepancy issue:

<img width="784" alt="screen shot 2017-05-11 at 10 40 00 pm" src="https://cloud.githubusercontent.com/assets/5607844/25980208/0e84ab98-369b-11e7-8e42-9735ffaad29f.png">

**Comment by nickidlugash** [comment]
[minimal style sheet](https://gist.github.com/nickidlugash/dcc431c8e6769aca9e801ca47370d960) (used in the screenshots above).

**Comment by nickidlugash** [comment]
Ref https://github.com/mapbox/mapbox-gl-js/issues/858 and https://github.com/mapbox/mapbox-gl-js/pull/860

**Comment by lbud** [comment]
Based on https://github.com/mapbox/mapbox-gl-js/pull/860#issuecomment-69130242 it seems like "should fix this issue's problem as well as we can right now" = this wasn't fully fixed. It _looks_ like the fix implemented was the middle option of https://github.com/mapbox/mapbox-gl-js/pull/860#issuecomment-68781678.

https://github.com/mapbox/mapbox-gl-js/commit/359a576471e20373333cae80a5eccda95ee4cdc2 was the fix implemented before, which was then ported to the shader in https://github.com/mapbox/mapbox-gl-js/commit/fcb9bcebae6e00f24ebfde8d7ec09adb95f6b1a3.

If I'm reading all this right, and we still want to close the remaining gap, we'd have to cause some line overlap for non-opaque lines by shifting the gap inward even more (the third option from the comment linked above). @nickidlugash @ansis — does that sound right, and does it sound like what we should do?

**Comment by ansis** [comment]
> If I'm reading all this right, and we still want to close the remaining gap, we'd have to cause some line overlap for non-opaque lines by shifting the gap inward even more (the third option from the comment linked above). @nickidlugash @ansis — does that sound right, and does it sound like what we should do?

Yeah, I think that sounds right.

**Comment by nickidlugash** [comment]
@lbud @ansis My 2c is that because we don't have a per-layer opacity option for lines, the usability of semi-transparent lines in conjunction with `line-gap-width` is very limited anyway, so I would prioritize what it looks like for opaque lines. But I could be biased towards my own use cases 😬

**Comment by ansis** [comment]
> @lbud @ansis My 2c is that because we don't have a per-layer opacity option for lines, the usability of semi-transparent lines in conjunction with line-gap-width is very limited anyway, so I would prioritize what it looks like for opaque lines.

Yep, I agree. I think this is what @lbud suggested, right?

**Comment by nickidlugash** [comment]
> does that sound right, and does it sound like what we should do?

Yep, I was confirming that I think we should go ahead and do this 👍
